homeassistant/components/habitica/calendar.py

```diff
@@ -0,0 +1,179 @@
+            if task["type"] == HabiticaTaskType.TODO
+            and not task["completed"]
+            and task.get("date")
+            and (start := to_date(task["date"]))
+            and start >= dt_util.now().date()
+        ]
+        events_sorted = sorted(
+            events,
+            key=lambda event: (
+                event.start,
+                self.coordinator.data.user["tasksOrder"]["todos"].index(event.uid),
+            ),
+        )
+
+        return events_sorted[0] if events_sorted else None
+
+    async def async_get_events(
+        self, hass: HomeAssistant, start_date: datetime, end_date: datetime
+    ) -> list[CalendarEvent]:
+        """Return calendar events within a datetime range."""
+
+        return [
+            CalendarEvent(
+                start=start,
+                end=start + timedelta(days=1),
+                summary=task["text"],
+                description=task["notes"],
+                uid=task["id"],
+            )
+            for task in self.coordinator.data.tasks
+            if task["type"] == HabiticaTaskType.TODO
+            and not task["completed"]
+            and task.get("date")
+            and (start := to_date(task["date"]))
+            and (start_date.date() <= start <= end_date.date())
+        ]
+
+
+class HabiticaDailiesCalendarEntity(HabiticaCalendarEntity):
+    """Habitica dailies calendar entity."""
+
+    entity_description = CalendarEntityDescription(
+        key=HabiticaCalendar.DAILIES,
+        translation_key=HabiticaCalendar.DAILIES,
+    )
+
+    @property
+    def event(self) -> CalendarEvent | None:
+        """Return the next upcoming event."""
+
+        today = datetime.combine(
+            dt_util.now().date(), time(), dt_util.DEFAULT_TIME_ZONE
+        )
+        events = [
+            CalendarEvent(
+                start=(start := next_recurrence.date()),
+                end=start + timedelta(days=1),
+                summary=task["text"],
+                description=task["notes"],
+                uid=task["id"],
+            )
+            for task in self.coordinator.data.tasks
+            if task["type"] == HabiticaTaskType.DAILY
+            and not task["completed"]
+            and task["everyX"]
+            if (next_recurrence := build_rrule(task).after(today, inc=True))
+        ]
+        events_sorted = sorted(
+            events,
+            key=lambda event: (
+                event.start,
+                self.coordinator.data.user["tasksOrder"]["dailys"].index(event.uid),
+            ),
+        )
+
+        return events_sorted[0] if events_sorted else None
+
+    async def async_get_events(
+        self, hass: HomeAssistant, start_date: datetime, end_date: datetime
+    ) -> list[CalendarEvent]:
+        """Return calendar events within a datetime range."""
+
+        today = to_date(self.coordinator.data.user["lastCron"])
+        # returns only todays and future dailies.
```

Comment:
This also is not meeting the get_event spec. get_events should work for any  start/end date range. A calendar can render events from the past as well, and triggers work based on past events.

Reply:
In Habitica the start of the day can be set to a different time (in my account it is 4:00 am for example. If at that time all dailies that were due have been completed the cron runs at 4 and resets all dailies, marking the start of the day. If there are incompleted dailies from the past day (called yesterdailies), the user gets a last chance to mark tasks as completed (in case the task was done but forgotten to mark as completed) and then the user presses the button 'start my day', which then marks the start of the day. That's what I want to reflect here. 
That's also why it doesn't make sense to render events from the past, dailies are reset every day.

Reply:
Also there is no data about past dailies. I could generate events based on the current recurrence settings but that would be a wild guess, because they can be changed anytime and I don't know if they were possibly different in the past

Reply:
Aren't the triggers based on the calendar entities on/off state? 
I checked the Todoist integration, it also provides a calendar for tasks with due dates. If a task is completed, it is not shown in the calendar view anymore. That's as far as I can see the only other integration that implements a calendar for tasks. Maybe in the future we could render completed tasks in the calendar in grey, that would be a cool feature, but for now I would prefer to do it like the Todoist integration does it already. 

I can't ignore the cron. I need it to determine on what day to render the tasks. That's the only way to know if the task is from yesterday or from today.

Reply:
I see logic in the "select best task" for skipping completed upcoming tasks. But in the calendar view, is that also happening? I saw a difference in the logic that gets the events https://github.com/home-assistant/core/blob/4cbac3a864e0724ad353aa3f4fc159cc8f402ae8/homeassistant/components/todoist/calendar.py#L664C15-L664C31

Regarding the cron are you saying the start date returned from the API is wrong?

Reply:
Ok, here is an example daily how it is returned by the Habitica API:
```
{
      "_id": "be011a27-58e0-4c93-9400-74176cb27da3",
      "frequency": "daily",
      "everyX": 1,
      "startDate": "2024-04-29T22:00:00.000Z",
      "repeat": {
        "m": true,
        "t": true,
        "w": true,
        "th": true,
        "f": true,
        "s": true,
        "su": true
      },
      "streak": 129,
      "daysOfMonth": [],
      "weeksOfMonth": [],
      "nextDue": [
        "2024-10-23T22:00:00.000Z",
        "2024-10-24T22:00:00.000Z",
        "2024-10-25T22:00:00.000Z",
        "2024-10-26T22:00:00.000Z",
        "2024-10-27T22:00:00.000Z",
        "2024-10-28T22:00:00.000Z"
      ],
      "yesterDaily": true,
      "completed": true,
      "collapseChecklist": false,
      "checklist": [],
      "type": "daily",
      "text": "Morgens Zähne putzen",
      "notes": "",
      "tags": [
        "b98f0391-ebc2-49a9-b308-fd03e3d77aad"
      ],
      "value": 79.8184124050229,
      "priority": 0.1,
      "attribute": "str",
      "challenge": {},
      "group": {
        "completedBy": {},
        "assignedUsers": []
      },
      "reminders": [],
      "byHabitica": false,
      "createdAt": "2024-04-30T21:47:15.656Z",
      "updatedAt": "2024-10-23T04:08:40.207Z",
      "userId": "4c4ca53f-c059-4ffa-966e-9d29dd405daf",
      "isDue": true,
      "id": "be011a27-58e0-4c93-9400-74176cb27da3"
    },
```

As you see, the tasks doesn't have a date (startDate is like dtstart for rrule). The task is either due or not due today (`isDue`). And to know what `today` means in habitica time, i need the date of the `lastCron`. So if the `lastCron` was today and the task `isDue`, I show it today (Real life today and Habitica today are the same). If the `lastCron` date is in the past, that means "Habitica today" diverges from real live today and I have to translate this to the real life time. 

The recurrences are like a preview of the task for the future. The API returns the next 6 recurrences in `nextDue`, but as there are only 6 dates returned, I have to calculate it myself with rrule so i can show recurrences between today and the end of the requested timespan, not only the next 6 days.